hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/CollectionPathNode.java

```diff
@@ -0,0 +1,281 @@
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.hql.internal.ast.tree;
+
+import java.util.List;
+
+import org.hibernate.QueryException;
+import org.hibernate.hql.internal.ast.HqlSqlWalker;
+import org.hibernate.hql.internal.ast.SqlASTFactory;
+import org.hibernate.internal.util.StringHelper;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.persister.collection.QueryableCollection;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.persister.entity.Joinable;
+import org.hibernate.persister.entity.PropertyMapping;
+import org.hibernate.type.CollectionType;
+import org.hibernate.type.CompositeType;
+import org.hibernate.type.EntityType;
+import org.hibernate.type.Type;
+
+import antlr.SemanticException;
+import antlr.collections.AST;
+
+/**
+ * @author Steve Ebersole
+ */
+public class CollectionPathNode extends SqlNode {
+	/**
+	 * Used to resolve the collection "owner key" columns
+ 	 */
+	private final FromElement ownerFromElement;
+
+	private final CollectionPersister collectionDescriptor;
+
+	private final String collectionPropertyName;
+	private final String collectionPropertyPath;
+	private final String collectionQueryPath;
+
+	private final HqlSqlWalker walker;
+
+
+	/**
+	 * Instantiate a `CollectionPathNode`
+	 *
+	 * @see #from(AST, AST, HqlSqlWalker)
+	 */
+	public CollectionPathNode(
+			FromElement ownerFromElement,
+			CollectionPersister collectionDescriptor,
+			String collectionPropertyName,
+			String collectionQueryPath,
+			String collectionPropertyPath,
+			HqlSqlWalker walker) {
+		this.ownerFromElement = ownerFromElement;
+		this.collectionDescriptor = collectionDescriptor;
+		this.collectionPropertyName = collectionPropertyName;
+		this.collectionQueryPath = collectionQueryPath;
+		this.collectionPropertyPath = collectionPropertyPath;
+		this.walker = walker;
+
+		walker.addQuerySpaces( collectionDescriptor.getCollectionSpaces() );
+
+		super.setType( SqlASTFactory.COLL_PATH );
+		super.setDataType( collectionDescriptor.getCollectionType() );
+		super.setText( collectionDescriptor.getRole() );
+	}
+
+	/**
+	 * Factory for `CollectionPathNode` instances
+	 *
+	 * @param qualifier The left-hand-side of a dot-ident node - may be null to indicate an ident arg
+	 * @param reference The right-hand-side of the dot-ident or the ident that is an unqualified reference
+	 */
+	public static CollectionPathNode from(
+			AST qualifier,
+			AST reference,
+			HqlSqlWalker walker) {
+
+		final String referenceName = reference.getText();
+		final String qualifierQueryPath = qualifier == null
+				? ""
+				: ( (FromReferenceNode) qualifier ).getPath();
+		final String referencePath = qualifier == null
+				? referenceName
+				: qualifierQueryPath + "." + reference;
+
```

Comment:
really a minor but probably we can initialize `qualifierQueryPath` and `referencePath` later
```
if ( qualifier == null ) {
   referencePath = referenceName;
....}else{
 final String qualifierQueryPath = ( (FromReferenceNode) qualifier ).getPath();
 referencePath = qualifierQueryPath + "." + reference;
}